feat(gux-pagination-beta): consolidate pagination options

joyshaffer opened the pull request:
Zeroheight documentation page with the updated and outdated pagination component: https://genesysds.zeroheight.com/styleguide/s/27050/p/62953d-pagination/b/15a8df

Figma design file: https://www.figma.com/file/qE1fsMVNIhP1y7oQyxuHCh/Pagination-Consolidation?node-id=0%3A1

This work is combining the "Full" and "Expanded" pagination options into "Advanced" and the "Small" pagination option is being renamed to "Simple."

I copied the stable/gux-pagination component into beta/gux-pagination-beta and added the updates in the beta component. It shows a lot of files changed because of the i18n translation files copied to the new component. Under gux-pagination-beta, the two pagination layout options are found under gux-pagination-advanced-beta and gux-pagination-simple-beta.

In stable/gux-pagination, I added some console warnings noting that if devs are using "Full" or "Expanded" pagination, they should start using the "Advanced" option in gux-pagination-beta, and if they are using "Small" pagination, they should start using the renamed "Simple" option in gux-pagination-beta.

daragh-king-genesys reviewed (approved):
I made some file changes to this PR
renamed files and reduced a bit of duplicate code

MattCheely commented on src/components/stable/gux-pagination/gux-pagination.tsx:
```diff
@@ -135,6 +137,22 @@ export class GuxPagination implements ComponentInterface {
+        'gux-pagination',
+        'Pagination "small" layout will be deprecated. Use gux-pagination-beta "simple" layout instead.'
+      );
+    }
```
I don't think we need to do this. We already have the component name lifecycle to handle letting people know when old versions are going away, and I'd like to reserve console messages for critical issues. If a lot of code is generating warnings, it means errors are likely to get ignored, and it becomes distracting noise when customers submit console logs to diagnose production issues.

joyshaffer commented:
> A few general comments
> 
> 1. If we are continuing to support the stable `gux-pagination` we should probably rename the new pagination component files `gux-pagination-beta`. I don't think the two components cannot have the same file names because the docs site does not support this. I think we ran into this with the `tabs-beta` component
> 2. We should consider whether the accessibility issues outlined in [COMUI-803](https://inindca.atlassian.net/browse/COMUI-803) can be added here. If not they will need to be added in a future PR. I don't really see any a11y regressions caused by the new pagination that weren't there previously
> 3. The space taken up by the pagination on the page varies depending on the width of the numbers. We should consider having a fixed width for each of the elements containing numbers. For example the pagination goes off the page when viewing the higher page numbers:
>    ![image](https://user-images.githubusercontent.com/82100934/164035043-ed2dadfb-a67b-4d1b-9269-6f5ebcf76750.png)
> 
> ![image](https://user-images.githubusercontent.com/82100934/164034960-be3df925-34d3-4cde-8329-f038b6888e5c.png)

1. I changed the naming for the component back to 'gux-pagination-beta'.
2. I could make these changes if needed, but also just considering keeping the scope of the PR to the addition of the beta pagination component.
3. I added an overflow:scroll property to the gux-pagination-container so the content won't overflow the page but also the elements won't overlap or be cut off because of set width properties.

joyshaffer commented:
CI build is failing but all tests are passing locally....

gavin-everett-genesys commented:
> CI build is failing but all tests are passing locally....

You will need to install the latest dependencies by using the command `npm ci` and then you will need to update your snapshots. The build should succeed then.

joyshaffer commented:
> > CI build is failing but all tests are passing locally....
> 
> You will need to install the latest dependencies by using the command `npm ci` and then you will need to update your snapshots. The build should succeed then.

Thanks that solved it!

katie-bobbe-genesys commented:
I'm seeing some styling issues with the pagination, which I think has to do with adding the overflow: scroll property. Perhaps try using [overflow-x](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-x) 
![image](https://user-images.githubusercontent.com/82100934/165160056-2c5e36d1-08d7-4864-a418-912638c6b9bf.png)

katie-bobbe-genesys commented:
I'm still seeing an issue where the pagination can get cut off depending on on the length of the numbers in the pagination list. Also I noticed that the focus styling on the buttons is cut off on the top and bottom.
[![Image from Gyazo](https://i.gyazo.com/b703591a4343bf3ac1abdc7122e48f32.gif)](https://gyazo.com/b703591a4343bf3ac1abdc7122e48f32)

joyshaffer commented:
> I'm still seeing an issue where the pagination can get cut off depending on on the length of the numbers in the pagination list. Also I noticed that the focus styling on the buttons is cut off on the top and bottom. [![Image from Gyazo](https://camo.githubusercontent.com/d612f7bc091b8ca07725f0ef785cdf84733c5b6471716c74e76721e65529d38d/68747470733a2f2f692e6779617a6f2e636f6d2f62373033353931613433343362663361633161626463373132326534386633322e676966)](https://gyazo.com/b703591a4343bf3ac1abdc7122e48f32)

The pagination in that scenario is able to be scrolled to. The scrolling behavior is needed so the pagination elements don't overlap each other. I hid the scrollbar styling as to not overlap and hide part of the pagination component.

daragh-king-genesys commented on src/components/beta/gux-pagination-beta/example.html:
```diff
@@ -0,0 +1,85 @@
+    <gux-pagination-beta
+      total-items="150"
+      items-per-page="25"
+      pages-unknown="true"
```
There is a general issue with this property in a few of the components. It is sometimes called `pages-unknown` and other times called `total-pages-unknown`.
Anyway I think we could have an API where we don't need an extra boolean flag (`pages-unknown` or `total-pages-unknown`) to handle that use case. `total-items` could be used for this too. Instead of defaulting it to `0` it could be defaulted to `undefined`. When it is `undefined` we can treat that as the unknown state.

joyshaffer replied:
I'm wondering if I can just omit the whole unknown total number of items/pages functionality all together. I _think_ that  `total-items` would always be whatever the total number of items are at that time the query happens. Because if `total-items` was undefined, then there would be no item count to display, so at that point it would make more sense to use the `cursor-pagination` component instead.

daragh-king-genesys commented:
> > I'm still seeing an issue where the pagination can get cut off depending on on the length of the numbers in the pagination list. Also I noticed that the focus styling on the buttons is cut off on the top and bottom. [![Image from Gyazo](https://camo.githubusercontent.com/d612f7bc091b8ca07725f0ef785cdf84733c5b6471716c74e76721e65529d38d/68747470733a2f2f692e6779617a6f2e636f6d2f62373033353931613433343362663361633161626463373132326534386633322e676966)](https://gyazo.com/b703591a4343bf3ac1abdc7122e48f32)
> 
> The pagination in that scenario is able to be scrolled to. The scrolling behavior is needed so the pagination elements don't overlap each other. I hid the scrollbar styling as to not overlap and hide part of the pagination component.

Can you confirm with the rest of your UX team that you are happy with this behaviour (scrollable region with no indication to the user that it is scrollable other than the possibility that an element appears cutoff/ to be missing). Just want confirmation as personally I think this is a bad idea.

daragh-king-genesys commented on src/components/beta/gux-pagination-beta/gux-pagination-beta.less:
```diff
@@ -0,0 +1,55 @@
+
+:host {
+  display: block;
+  border-top: 1px solid @gux-grey-60;
```
Linked to the scroll comments this is where I think overflow should be added and a (styled) scrollbar would appear under the control when needed.

joyshaffer replied:
We do not want a scroll bar visible with the pagination component. Also, if the viewport is designed within a smaller region, the simple pagination or cursor pagination should be used instead.

daragh-king-genesys replied:
👍

katie-bobbe-genesys commented:
It seems like there are no breaking API changes between the `gux-pagination` and `gux-pagination-beta` components. The addition of `gux-pagiation-beta` may be confusing for consumers of our component library, and it would be best to avoid unnecessary duplication. I propose that instead of making a new `gux-pagination-beta` component, we add these changes to the `gux-pagination` component.

layout property:
1. `full`/ default: Continue to support, but add warning to console that tells users that this layout will be deprecated in future release, change to use `advanced` layout instead
2. `expanded`: Continue to support but add warning to console that tells users that this layout will be deprecated in future release, change to use `advanced` layout instead
3. `small`/`simple`: both `small` and `simple` will map to the same layout, but add warning to console that tells users that `simple` prop name will be deprecated, and to use `small` instead
4. `advanced`: in a future release, change the `default` layout to be `advanced`

joyshaffer commented:
@katie-bobbe-genesys In previous discussions, it was decided that the best option would be to add the beta component as it would allow for a structured migration for dev teams (see comment on [ticket](https://inindca.atlassian.net/browse/COMUI-961)). Also, the console warnings are likely to be ignored and distracting from other issues.

daragh-king-genesys commented:
> @katie-bobbe-genesys In previous discussions, it was decided that the best option would be to add the beta component as it would allow for a structured migration for dev teams (see comment on [ticket](https://inindca.atlassian.net/browse/COMUI-961)). Also, the console warnings are likely to be ignored and distracting from other issues.

I know when I give my opinion earlier that we should go with a beta component I did not think this amount of duplicate code would be needed. Up to know I was sticking with it because of the API change that has just been removed. I think we should reevaluate how we are doing this.

My suggestion is to implement the suggestions @katie-bobbe-genesys made above without the warning to `gux-pagination`. Then we can create a `gux-pagination-beta` component that is a simple shell component that calls `gux-pagination`. the `layout` prop of `gux-pagination-beta` will only have the two new layouts as valid options.

joyshaffer commented:
If this is the route to go in, I was wondering how migrating in the future to just the `advanced` and `simple` layouts would work if `gux-pagination-beta` calls `gux-pagination`? If the advanced layout is just in the beta component, wouldn't the code needed for that layout have to be added to `gux-pagination-beta`? Just want to make sure I understand before starting the rework.

daragh-king-genesys commented:
> If this is the route to go in, I was wondering how migrating in the future to just the `advanced` and `simple` layouts would work if `gux-pagination-beta` calls `gux-pagination`? If the advanced layout is just in the beta component, wouldn't the code needed for that layout have to be added to `gux-pagination-beta`? Just want to make sure I understand before starting the rework.

This would be how i approached it.
v2 - `gux-pagination` (full, expanded, small)

v3 - `gux-pagination` (full, expanded, small, advanced, simple)
v3 - `gux-pagination-beta` (advanced, simple) - `gux-pagination-beta` is just a wrapper that calls `gux-pagination`.

v4 - `gux-pagination` (advanced, simple) - the code for full, expanded, small will be removed.
v4 - `gux-pagination-legacy` (full, expanded, small, advanced, simple) - This is a full copy of 'gux-pagination (v3)' and there would be no shared code with the stable component.
v4 - `gux-pagination-beta` would just be deleted

MattCheely commented:
Ok, so coming back from time off and various 🔥s to look at this, I see two things we need to resolve:

- The migration path
- Behavior at smaller sizes

**Migration**

On the migration front, I don't think we need to make significant changes. The current strategy may be more aggressive than necessary in hindsight, but it's also not going to break anyone.  On the duplicate code front, I don't see that causing significant problems. The older component is only going to get bug fixes at this point, but the newer one may need to evolve in more drastic ways, so having it use separate code paths may work in our favor. We may have to fix bugs in two places, but I don't think the pagination component has enough complexity to make that much of a burden. There's a lot less that can go wrong with it compared to more complex components.

**Small layouts**

On the other hand, I think the overflow/scroll behavior at small screen sizes is a non-starter.  Even if UX were to sign off on it, I'm confident we'd get a bug report on it as internal QA or a customer encounters it.  We need to go back to the design team and ask about the expected behavior at smaller sizes.  My recommendation would be to determine a reasonable  minimum size for the advanced layout, one that accomodates either 3 or 4 digit pages, and use DOM measurements and a `ResizeObserver` to switch to the `simple` layout if the component is smaller than that width.

@joyshaffer we can discuss this with the design team at the next review meeting, or I'm happy to have a conversation prior to that if needed.

daragh-king-genesys commented:
With regard to the migration and code duplication I agree there are pros and cons but I think we are much more likely to be duplicating fixes for the next year than the new design to significantly evolve. So the likely con of the overhead to duplicating fixes may not be worth the unlikely pro of easier divergence over the next year.

MattCheely commented:
I think if we set up a ResizeObserver to switch to the small layout if the component is narrower than 810px it's enough for us to move on with this one for now.  Daragah and I will just have to disagree about the duplicate code for a bit, or we can refactor it later.

joyshaffer commented:
Resize behavior has been added

katie-bobbe-genesys commented on src/components/beta/gux-pagination-beta/gux-pagination-beta.tsx:
```diff
@@ -57,6 +60,12 @@ export class GuxPaginationBeta implements ComponentInterface {
+   * Previous pagination layout - for resize observer
+   */
+  @State()
+  previousLayout: GuxPaginationLayoutBeta;
```
Instead of using `previousLayout`, it might make sense to just reference the `layout` prop for the initial layout type. Instead of setting the `layout` prop to simple or advanced based on the width, you could make a new `State` called something like `displayedLayout`. So, `layout` is the pagination layout was initially set to, and `displayedLayout` will change based on the breakpoint. 

I think the pagination component is looking really good, and the addition of this resize observer is a great feature

MattCheely commented on src/components/beta/gux-pagination-beta/gux-pagination-beta.tsx:
```diff
@@ -0,0 +1,210 @@
+        '.gux-pagination-container'
+      );
+      const paginationContainerWidth = el.clientWidth;
+      if (paginationContainerWidth < 810) {
```
This 810 is a bit mysterious in this context for a maintainer.  Best practice with these sort of values is to define and document them at the top of the file. Something like:

```typescript
/**
 * The minimum size in px at which the advanced layout will display. Below
 * this size we switch to the simple layout.
 * This was chosen based on the width required for 3-digit page numbers, so 
 * page counts that go into 4 digits could overflow at small sizes.
 * /
const MIN_ADVANCED_SIZE = 810;
```
That way if it gets re-used later in another context, it doesn't end up defined twice, and maintainers can understand when/why we'd want to change it. 

Other than that, I think this looks pretty good. Let me know when this change is in and we'll get it merged.

joyshaffer replied:
@MattCheely MIN_ADVANCED_SIZE constant has been added
